dm/master/openapi_test.go

```diff
@@ -207,3 +218,177 @@ func (t *openAPISuite) TestSourceAPI(c *check.C) {
+	result4 := testutil.NewRequest().Patch(startRelayURL).WithJsonBody(openAPIStartRelayReq).Go(t.testT, s.echo)
+	// check http status code
+	c.Assert(result4.Code(), check.Equals, http.StatusOK)
+
+	// get source status again, relay status should not be nil
+	result5 := testutil.NewRequest().Get(source1StatusURL).Go(t.testT, s.echo)
+	c.Assert(result5.Code(), check.Equals, http.StatusOK)
+	var resultSourceStatus2 openapi.SourceStatus
+	err = result5.UnmarshalBodyToObject(&resultSourceStatus2)
+	c.Assert(err, check.IsNil)
+	c.Assert(resultSourceStatus2.RelayStatus.Stage, check.Equals, pb.Stage_Running.String())
+
+	// test stop relay
+	stopRelayURL := fmt.Sprintf("%s/stop-relay", source1URL)
+	workerNameReq := openapi.WorkerNameRequest{WorkerName: workerName1}
+	result6 := testutil.NewRequest().Patch(stopRelayURL).WithJsonBody(workerNameReq).Go(t.testT, s.echo)
+	c.Assert(result6.Code(), check.Equals, http.StatusOK)
+}
+
+func mockWorkerClientForOpenAPI(
+	mockWorkerClient *pbmock.MockWorkerClient, taskName, sourceName, workerName string, masterReq interface{}) {
+	var expect pb.Stage
+	switch req := masterReq.(type) {
+	case *pb.StartTaskRequest, *pb.UpdateTaskRequest:
+		expect = pb.Stage_Running
+	case *pb.OperateTaskRequest:
+		switch req.Op {
+		case pb.TaskOp_Resume:
+			expect = pb.Stage_Running
+		case pb.TaskOp_Pause:
+			expect = pb.Stage_Paused
+		case pb.TaskOp_Stop:
+			expect = pb.Stage_Stopped
+		}
+	case *pb.OperateWorkerRelayRequest:
+		switch req.Op {
+		case pb.RelayOp_ResumeRelay:
+			expect = pb.Stage_Running
+		case pb.RelayOp_PauseRelay:
+			expect = pb.Stage_Paused
+		case pb.RelayOp_StopRelay:
+			expect = pb.Stage_Stopped
+		}
+	default:
+		expect = pb.Stage_Running
+	}
+	queryResp := &pb.QueryStatusResponse{
+		Result: true,
+		SourceStatus: &pb.SourceStatus{
+			Worker:      workerName,
+			Source:      sourceName,
+			RelayStatus: &pb.RelayStatus{Stage: expect},
+		},
+		SubTaskStatus: []*pb.SubTaskStatus{
+			{
+				Stage: expect,
+				Name:  taskName,
+				Status: &pb.SubTaskStatus_Sync{
+					Sync: &pb.SyncStatus{
+						TotalEvents:         0,
+						TotalTps:            0,
+						RecentTps:           0,
+						MasterBinlog:        "",
+						MasterBinlogGtid:    "",
+						SyncerBinlog:        "",
+						SyncerBinlogGtid:    "",
+						BlockingDDLs:        nil,
+						UnresolvedGroups:    nil,
+						Synced:              false,
+						BinlogType:          "",
+						SecondsBehindMaster: 0,
+					},
+				},
+			},
+		},
+	}
+
+	switch masterReq.(type) {
+	case *pb.StartTaskRequest, *pb.UpdateTaskRequest, *pb.OperateTaskRequest:
+		queryResp.SubTaskStatus = []*pb.SubTaskStatus{{}}
+		if expect == pb.Stage_Stopped {
+			queryResp.SubTaskStatus[0].Status = &pb.SubTaskStatus_Msg{
+				Msg: fmt.Sprintf("no sub task with name %s has started", taskName),
+			}
+		} else {
+			queryResp.SubTaskStatus[0].Name = taskName
+			queryResp.SubTaskStatus[0].Stage = expect
+		}
+	case *pb.OperateWorkerRelayRequest:
```

Comment:
why this type is repeated assignment apart from line 341

Reply:
mybad, i copy mock from `mockRevelantWorkerClient` that maybe cause some misunderstanding, i fix this mock in https://github.com/pingcap/dm/pull/2083/commits/4df4e65114af4b1dae18386b16d642da1fa131e1